### DISCUSSION TITLE
in-line casting using generics

### DISCUSSION THREAD

**Discussion by Kr00l** [twinbasic/twinbasic#705] [Ideas]
Hello,

I was curios about the in-line casting mentioned in #691 

I tried in tB to achieve this using generics. My idea was:
```
Class TCast(Of T)
Attribute VB_PredeclaredId = True

Public Function CObj(ByRef Obj As Object) As T
Attribute CObj.VB_UserMemId = 0
Return Obj
End Function
    
End Class
```

It's predeclared (global instance) and has a default member. So I was expecting to have below code would work and debug prints True.
```
Public Sub Main()
    
    Dim This As TestClass = New TestClass
    
    Dim Temp As ITestClass
    Set Temp = This
    
    Debug.Print ObjPtr(TCast(Of ITestClass)(This)) = ObjPtr(Temp)
    
End Sub
```

The testing classes:
```
Class ITestClass
End Class

Class TestClass
Implements ITestClass
End Class
```

But it gives error saying unrecognized symbol TCast.

New Result as of discussion:
```
Public Function TCast(Of TOut, TIn)(ByRef Expression As TIn) As TOut
Return Expression
End Function
```

**Comment by bclothier** [comment]
How can you predeclare an open generic type? I think the problem is that you must close the generic type in order to access the instance, but we are not closing it when we decorate it with the predeclaredId attribute. 

Try using an instance so that you can close it. 

```
Dim x As TCast(Of ITestClass)
Set Temp = x.Cast(Test)
```

**Reply by WaynePhillipsEA** [reply]
You're right there.  Though I have no idea if the compiler will allow that through... that's definitely in untested territory!

**Reply by bclothier** [reply]
One more alternative. Instead of making the class generic, make the method generic. That way, the class does not need to be closed but you can close it at the invocation of the `CObj` method.

**Reply by Kr00l** [reply]
I tried with a function at first. But how can a function be generic and accept an param at the same time?

**Reply by bclothier** [reply]
Haven't tried, but I think in VB.NET the syntax goes like `Function CObj(Of T)(ByRef Obj As T)`

(caveat: I don't do VB.NET so it may be wrong)

**Reply by WaynePhillipsEA** [reply]
Yes, that's the correct syntax

**Reply by Kr00l** [reply]
Thanks. I tried this
```
Function TCast(Of T, Obj As Object) As T
Return Obj
End Function
```

So below would work? (Need to test later)
```
Function TCast(Of T)(Obj As Object) As T
Return Obj
End Function
```

**Comment by Kr00l** [comment]
Great, it works. Thanks
I ended up with the below universal function.

```
Public Function TCast(Of T)(ByRef Expression As T) As T
Return Expression
End Function
```

So it can be used for all kinds of thing, also e.g.
`Debug.Print TCast(Of Date)("2021-01-01")`

**Reply by WaynePhillipsEA** [reply]
Yes, that's pretty cool as you can use it to inline-cast object/interfaces too.

**Reply by bclothier** [reply]
Hmm, Is that really different from doing:

```
Dim d As Date = "2021-01-01"
```

Because your function is defined to take `Expression As T` as the input, the implicit conversion will already have happened before entering the function. To defer the conversion, I think you need it to be `Expression As Variant` but keep in mind that an implicit conversion to `Variant` will still happen.

To completely eliminate the implicit conversions, I think it may have to be:
```
Public Function Cast(Of TIn, TOut)(ByRef Expression As TIn) As TOut
```

The compiler should have no problem inferring the `TIn` but I think `TOut` still has to be supplied, unless you use it in an assignment like:

```
Dim d As Date = Cast("2021-02-01")
```

Not sure how much inference the compiler can do right now.

I also imagine you can also write a `TryCast` like so:
```
Public Function TryCast(Of TIn, TOut)(ByRef Expression As TIn, OutResult As TOut) As Boolean
```

which you can then provide the error handling and return false if it cannot be cast successfully.

**Reply by WaynePhillipsEA** [reply]
I assumed the idea for this was to do something like `Cast(Of Access.Application)(MyObject).OpenDatabase(...)`... i.e. generically cast to a different interface without needing to define an extra variable, but I admit that I may have read the intention wrong here.

**Reply by WaynePhillipsEA** [reply]
> Not sure how much inference the compiler can do right now.

It won't look at the return side of the expression for inference purposes.

Note that you can define your example slightly differently so that you can still gain inference of the TIn:
`Public Function Cast(Of TOut, TIn)(ByRef Expression As TIn) As TOut`
then it will let you just supply the first generic specifier, and the second will be inferred... e.g. `Cast(Of VBA.Collection)(obj)`

**Reply by Kr00l** [reply]
> I assumed the idea for this was to do something like `Cast(Of Access.Application)(MyObject).OpenDatabase(...)`... i.e. generically cast to a different interface without needing to define an extra variable, but I admit that I may have read the intention wrong here.

Yes that was the intention.

**Reply by bclothier** [reply]
In this case, the `Expression` should be just an `Object`, not `T`, then we don't need 2 different type parameters.

**Reply by WaynePhillipsEA** [reply]
But not all interfaces support IDispatch, so using a generic type is more universal.

**Reply by bclothier** [reply]
Would using `stdole.IUnknown` have worked? Otherwise, we need to make a distinction between a COM interface and a data type. 

Edit: I'm probably forgetting something - we have an attribute for that case already?

**Reply by WaynePhillipsEA** [reply]
Yes IUnknown would work, but due to the rules of IUnknown having to be a unique "identity" interface, it will cause an extra QueryInterface call at the callsite.

**Reply by WaynePhillipsEA** [reply]
We should be able guard the generic type being an object using the `[CompileIf]` attribute, yes

**Reply by Kr00l** [reply]
Ah yes, it's like a constraint for the generics this `[CompileIf]`

**Reply by WaynePhillipsEA** [reply]
Yes, but actually we can't do `[ CompileIf (IsObject(Of TIn)) ]` currently, because it won't _see_ the `TIn` type defined as part of the function... `[CompileIf]` can only evaluate types defined at the wider component scope (i.e. the module in this case... which obviously can't have generic types)

**Reply by WaynePhillipsEA** [reply]
Actually, we shouldn't need to explicitly guard against it, since setting an object variable to a non-object expression should be a compile error already.   For example, tB should not be allowing `Return Expression` if Expression is a Long and the return type is a VBA.Collection.  It actually doesn't allow this, but the compile error isn't triggered until codegen time, which is a compiler error I'll get fixed tomorrow.

So I would say `Public Function Cast(Of TOut, TIn)(ByRef Expression As TIn) As TOut` is the correct way to do this.

**Reply by mansellan** [reply]
Hmm, just thinking, can't we do this?

`Public Function Cast(Of TOut)(ByRef Expression As IDispatch) As TOut`? (plus another overload for `IUnknown` if that works too)

**Reply by mansellan** [reply]
Also, it occurs to me that such universal functions as these could really benefit from being part of a community "standard library". Should we start a GH repo for that?

**Reply by bclothier** [reply]
> Hmm, just thinking, can't we do this?

Hint: `Object` == `IDispatch`. ;-) See earlier remarks.

**Reply by bclothier** [reply]
> Should we start a GH repo for that? 

I certainly agree for other cases but for this particular case, I am wondering if it's best done by the compiler.

**Reply by Kr00l** [reply]
> > Should we start a GH repo for that?
> 
> I certainly agree for other cases but for this particular case, I am wondering if it's best done by the compiler.

What you mean by the compiler? If it's a source code snippet it's always by the compiler(?)

**Reply by bclothier** [reply]
I mean inlined by compiler, similar to how VBA compiler optimizes the calls for some of the conversion functions. That never happens for a user-defined functions.

**Reply by WaynePhillipsEA** [reply]
> I mean inlined by compiler, similar to how VBA compiler optimizes the calls for some of the conversion functions. That never happens for a user-defined functions.

... yet 😜

**Reply by WaynePhillipsEA** [reply]
FYI, as of 0.14.6, `Set Object = <non-object>` and variations like trying to `Return <non-object>` to an Object returning function will now cause a proper compile-error, rather than the codegen error given in previous versions.
